File: QMB-Spesialpensum/RBM/WaveFunctions/.ipynb_checkpoints/simplegaussian-checkpoint.py
import numpy as np
from particles import Particles   
from numba import int32, float64
from numba.experimental import jitclass
from WaveFunctions.wavefunction import WaveFunction

spec = [
    ("_alpha", float64),
    ("_HOs", float64[:])
]

@jitclass(spec)
class SimpleGaussian(WaveFunction):
    def __init__(self, alpha, HOshape):
        self._alpha = alpha
        self._HOs = HOshape
    
    def kinetic(self, particles):
        return self._alpha * np.sum(self._HOs - 2 * self._alpha * np.power(particles.getPositions() * self._HOs, 2))
    
    def QF(self, particles, p1):
        return -4 * self._alpha * particles.getPos(p1) * self._HOs
    
    def ratio(self, particles, p1, oldPos):
        """
        Old wavefunc value term: exp(-alpha * old_r2)
        New wavefunc value term: exp(-alpha * r2)
        All other terms are the same and cancel. Since they are both exponentials, we can subtract the exponents
        """
        r2 = np.sum(particles.getPos(p1) * particles.getPos(p1) * self._HOs)
        old_r2 = np.sum(oldPos * oldPos * self._HOs)
        return np.exp(-self._alpha * r2 + self._alpha * old_r2)
    
    def paramDer(self, particles):
        positions = particles.getPositions()
        num, _ = particles.getSize()
        return -np.sum(positions * positions * self._HOs) / num
    
    def updateParam(self, alpha):
        self._alpha = alpha

File: QMB-Spesialpensum/Julia VMC/Wavefunctions/.ipynb_checkpoints/simplegaussian-checkpoint.jl
import LinearAlgebra as la
import StaticArrays as sa

struct SimpleGaussian{V} <: WaveFunction
    alpha::Float64
    HOshape::V
    HOshape2::V
    function SimpleGaussian(alpha, HOshape)
        HOshape = sa.SVector{size(HOshape,1), Float64}(HOshape)
        return new{typeof(HOshape)}(alpha, HOshape, HOshape.^2)
    end
end


function ratio(particles, p1, old_pos, wf::SimpleGaussian)::Float64
    """
    Old wavefunc value term: exp(-alpha * old_r2)
    New wavefunc value term: exp(-alpha * r2)
    All other terms are the same and cancel. Since they are both exponentials, we can subtract the exponents
    """
    temp_vec = particles.temp_vec
    temp_vec .= old_pos.^2
    temp_vec .= temp_vec .- particles.positions[p1].^2
    temp_vec .= temp_vec .* wf.HOshape
    return exp(wf.alpha * sum(temp_vec))
end

function kinetic(particles, wf::SimpleGaussian)::Float64
    temp_vec = zero(particles.temp_vec) #maybe there is a faster way? why does this not allocate??
    for pos in particles.positions
        temp_vec .+= pos.^2
    end
    temp_vec .= temp_vec .* wf.HOshape2
    return wf.alpha * (sum(wf.HOshape) * particles.num - 2.0 * wf.alpha * sum(temp_vec))
end

function QF(particles, p1, wf::SimpleGaussian) #maybe faster to put return type here?
    particles.temp_vec .= -4 * wf.alpha .* particles.positions[p1] .* wf.HOshape
    return particles.temp_vec
end

function paramDer(particles, wf::SimpleGaussian)::Float64
    temp_vec = zero(particles.temp_vec) #maybe there is a faster way? why does this not allocate??
    for pos in particles.positions
        temp_vec .+= pos.^2
    end
    temp_vec .= temp_vec .* wf.HOshape
    return -sum(temp_vec) / particles.num
end

function applyGradient(wf::SimpleGaussian, grad)
    return SimpleGaussian(wf.alpha - grad, wf.HOshape)
end